haystack/core/component/sockets.py

```diff
@@ -1,57 +1,106 @@
 # SPDX-FileCopyrightText: 2022-present deepset GmbH <info@deepset.ai>
 #
 # SPDX-License-Identifier: Apache-2.0
+
 import logging
-from dataclasses import dataclass, field
-from typing import Any, List, Type, get_args
+from typing import Dict, Type, Union
+
+from haystack.core.type_utils import _type_name
 
-from haystack.core.component.types import HAYSTACK_VARIADIC_ANNOTATION
+from .types import InputSocket, OutputSocket
 
 logger = logging.getLogger(__name__)
 
+SocketsType = Union[Type[InputSocket], Type[OutputSocket]]
+
+
+class Sockets:
+    """
+    This class is used to represent the inputs or outputs of a `Component`.
+    Depending on the type passed to the constructor, it will represent either the inputs or the outputs of
+    the `Component`.
+
+    Usage:
+    ```python
+    from haystack.components.builders.prompt_builder import PromptBuilder
+
+    prompt_template = \"""
+    Given these documents, answer the question.\nDocuments:
+    {% for doc in documents %}
+        {{ doc.content }}
+    {% endfor %}
+
+    \nQuestion: {{question}}
+    \nAnswer:
+    \"""
 
-class _empty:
-    """Custom object for marking InputSocket.default_value as not set."""
+    prompt_builder = PromptBuilder(template=prompt_template)
+    sockets = {"question": InputSocket("question", Any), "documents": InputSocket("documents", Any)}
+    inputs = Sockets(component=prompt_builder, sockets=sockets, sockets_type=InputSocket)
+    inputs
+    >>> PromptBuilder inputs:
+    >>>   - question: Any
+    >>>   - documents: Any
 
+    inputs.question
+    >>> InputSocket(name='question', type=typing.Any, default_value=<class 'haystack.core.component.types._empty'>, is_variadic=False, senders=[])
+    ```
+    """
 
-@dataclass
-class InputSocket:
-    name: str
-    type: Type
-    default_value: Any = _empty
-    is_variadic: bool = field(init=False)
-    senders: List[str] = field(default_factory=list)
+    # We're using a forward declaration here to avoid a circular import.
+    def __init__(
+        self,
+        component: "Component",  # type: ignore[name-defined] # noqa: F821
+        sockets: Dict[str, SocketsType],
+        sockets_type: SocketsType,
+    ):
+        """
+        Create a new Sockets object.
+        We don't do any enforcement on the types of the sockets here, the `sockets_type` is only used for
+        the `__repr__` method.
+        We could do without it and use the type of a random value in the `sockets` dict, but that wouldn't
+        work for components that have no sockets at all. Either input or output.
+        """
+        self._sockets_type = sockets_type
+        self._component = component
+        self._sockets = sockets
```

Comment:
Since this is private (and I understand in this PR we access it from the caller only as a temporary solution) can we call it something closer to the implementation detail, like `sefl._dict` or `self._sockets_names`? Otherwise it's confusing having `Sockets._sockets`.